File: jacfarm-frontend/src/api/logs.ts
import { API_URI, PAGE_LIMIT, STATUS_OK } from "./api";
import type { GetLogLevelsResponse, GetLogsResponse, GetModulesResponse } from "./responses";


export const getLogs = async (page: number, moduleId: number | null, exploitId: string | null, logLevelId: number | null) => {
    if (moduleId == null) moduleId = 0;
    if (exploitId == null) exploitId = "";
    if (logLevelId == null) logLevelId = 0;

    const res = await fetch(`${API_URI}/api/v1/logs?page=${page}&limit=${PAGE_LIMIT}&module_id=${moduleId}&exploit_id=${exploitId}&log_level_id=${logLevelId}`, {credentials: "include"});
    const data: GetLogsResponse = await res.json();

    if (data.status != STATUS_OK) {
        throw Error(data.error);
    }

    return {
        logs: data.logs,
        count: data.count
    }
}

export const getModules = async () => {
    const res = await fetch(`${API_URI}/api/v1/logs/modules`, {credentials: "include"});
    const data: GetModulesResponse = await res.json();

    if (data.status != STATUS_OK) {
        throw Error(data.error);
    }

    return {
        modules: data.modules,
        count: data.count
    }
}

export const getLogLevels = async () => {
    const res = await fetch(`${API_URI}/api/v1/logs/levels`, {credentials: "include"});
    const data: GetLogLevelsResponse = await res.json();

    if (data.status != STATUS_OK) {
        throw Error(data.error);
    }

    return {
        log_levels: data.log_levels,
        count: data.count
    }
}

File: jacfarm-api/internal/http/handlers/logs.go
package handlers

import (
	"JacFARM/internal/http/dto"

	"github.com/gofiber/fiber/v3"
)

// ListLogs godoc
// @Summary Возврат полной информации о логах
// @Param limit query int false "Количество логов"
// @Param page query int false "Страница логов"
// @Param exploit_id query int false "Идентификатор эксплойта"
// @Param module_id query int false "Идентификатор модуля"
// @Param log_level_id query int false "Идентификатор уровня логирования"
// @Produce json
// @Success 200 {object} dto.ListLogsResponse
// @Failure 400 {object} dto.Response "Неверный запрос"
// @Failure 401 {object} dto.Response "Ошибка авторизации"
// @Failure 500 {object} dto.Response "Внутренняя ошибка"
// @Router /api/v1/logs [get]
// @Tags Logs
// @Security BasicAuth
func (h *Handlers) ListLogs() func(c fiber.Ctx) error {
	return func(c fiber.Ctx) error {
		filter, err := dto.MapQueryToListLogsFilter(c.Queries())
		if err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(dto.Error(err.Error()))
		}

		logs, count, err := h.service.ListLogs(c.RequestCtx(), filter)
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(dto.ErrInternal)
		}

		return c.JSON(dto.ListLogsResponse{
			Response: dto.OK(),
			Logs:     logs,
			Count:    count,
		})
	}
}

// ListModules godoc
// @Summary Возврат всех модулей фермы
// @Produce json
// @Success 200 {object} dto.ListModulesResponse
// @Failure 400 {object} dto.Response "Неверный запрос"
// @Failure 401 {object} dto.Response "Ошибка авторизации"
// @Failure 500 {object} dto.Response "Внутренняя ошибка"
// @Router /api/v1/logs/modules [get]
// @Tags Logs
// @Security BasicAuth
func (h *Handlers) ListModules() func(c fiber.Ctx) error {
	return func(c fiber.Ctx) error {
		modules, count, err := h.service.ListModules(c.RequestCtx())
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(dto.ErrInternal)
		}

		return c.JSON(dto.ListModulesResponse{
			Response: dto.OK(),
			Modules:  modules,
			Count:    count,
		})
	}
}

// ListLogLevels godoc
// @Summary Возврат всех уровней логирования
// @Produce json
// @Success 200 {object} dto.ListLogLevelsResponse
// @Failure 400 {object} dto.Response "Неверный запрос"
// @Failure 401 {object} dto.Response "Ошибка авторизации"
// @Failure 500 {object} dto.Response "Внутренняя ошибка"
// @Router /api/v1/logs/levels [get]
// @Tags Logs
// @Security BasicAuth
func (h *Handlers) ListLogLevels() func(c fiber.Ctx) error {
	return func(c fiber.Ctx) error {
		logLevels, count, err := h.service.ListLogLevel(c.RequestCtx())
		if err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(dto.ErrInternal)
		}

		return c.JSON(dto.ListLogLevelsResponse{
			Response:  dto.OK(),
			LogLevels: logLevels,
			Count:     count,
		})
	}
}
